### DISCUSSION TITLE
Cannot specify `Depends` for type Response, yet docs imply you can?

### DISCUSSION THREAD

**Discussion by holyoaks** [fastapi/fastapi#9897] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated
import arrow
from fastapi import FastAPI, Response

app = FastAPI()

def set_no_cache_headers(response: Response) -> Response:
    response.headers["Expires"] = arrow.utcnow().format(arrow.FORMAT_RFC1123)
    response.headers["Cache-Control"] = "max-age=0, no-cache, no-store, must-revalidate, private"
    return response


class NoCacheResponse:
    def __init__(self, header_directives: List[str]) -> None:
        self.header_directives = header_directives

    def __call__(self, response: Response) -> None:
        if "no_cache" in self.header_directives:
            set_no_cache_headers(response=response)


NO_CACHE_RESPONSE = NoCacheResponse(["no_cache"])

@app.get("/testing/")
def testing_response(
    response: Annotated[Response, Depends(NO_CACHE_RESPONSE)],
):
    return {}
```

### Description

Previously (up to FastAPI 0.89.1, which is the last version I used before trying to update), the code above worked just fine. I use this to add caching headers to my GET responses (as certain network devices will cache unless specifically told not to). I built this as the docs [here](https://fastapi.tiangolo.com/advanced/response-headers/#use-a-response-parameter) specify that `You can also declare the Response parameter in dependencies, and set headers (and cookies) in them.`. Again, it worked fine until at least 0.89.1.

Now, FastAPI specifically block this functionality. Lines 383-387 of `fastapi/dependencies/utils.py`:

```
if lenient_issubclass(
        type_annotation,
        (Request, WebSocket, HTTPConnection, Response, BackgroundTasks, SecurityScopes),
    ):
        assert depends is None, f"Cannot specify `Depends` for type {type_annotation!r}"
```

 checks to see if the parameter is of certain types, and immediately fails indicating `AssertionError: Cannot specify `Depends` for type <class 'starlette.responses.Response'>`. 

My question is this: Which is correct, the code or the docs? If I'm simply misunderstanding how to `declare the Response parameter in dependencies`, please help me understand how best to accomplish what I used to be able to. Thanks!

### Operating System

Linux

### FastAPI Version

0.100.0

### Python Version

3.11

**Comment by Kludex** [comment]
Can you reproduce with fastapi 0.99.0?

**Reply by holyoaks** [reply]
I can. Exact same behavior, in the exact same `lenient_issubclass` method

**Reply by holyoaks** [reply]
@Kludex Should I create an issue from this discussion? Don't want to see this fall through the cracks

**Reply by Kludex** [reply]
https://github.com/tiangolo/fastapi/issues/10127

**Comment by Kludex** [comment, marked as answer]
This is a bug.

For the time being, you can do the following as a workaround:

```py
from typing import Annotated
from fastapi import FastAPI, Response, Depends

app = FastAPI()

def set_no_cache_headers(response: Response) -> Response:
    response.headers["Potato"] = "potato"
    return response

@app.get("/")
def testing_response(response = Depends(set_no_cache_headers)):
    return {}
```
